src/storage/src/render/upsert.rs

```diff
@@ -103,11 +106,34 @@ impl<H: Digest> Hasher for DigestHasher<H> {
     }
 }
 
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+struct UpsertState {
+    data: Result<Row, UpsertError>,
+    order: Option<MzOffset>,
+}
+
+/// Gets the field to compare the records by and returns the value
+pub(crate) fn get_order(row: &Row, order_by_index: Option<usize>) -> Option<MzOffset> {
+    order_by_index.map(|index| {
+        let datum = row
+            .iter()
+            .nth(index)
+            .expect("Expected a valid order by index");
+        match datum {
+            Datum::Timestamp(ts) => MzOffset {
+                offset: ts.timestamp_millis() as u64,
```

Comment:
I don't think this `as` cast has the right semantics, as—I think—it will sort negative timestamps (which are legit, and represent dates before Jan 1 1970) _after_ the maximum positive timestamp. If this must go into an `MzOffset`, I think you can make this work with some logic like

```rust
let mut ts = u64::reinterpret_cast(ts.timestamp_millis());
if ts >= 9223372036854775808 {
  ts -= 9223372036854775808;
} else {
  ts += 9223372036854775808;
}
```

but that's pretty headscratching. If there's a way to make the ordering take a `Datum` directly instead (maybe going back to the `O: Ord` generic that the upsert logic took previously?) that's probably best. Would mean the upsert envelope had sensible behavior regardless of the ordering column—and you could remove mention of the restriction from the `UpsertEnvelope` docstring.